Fail at build time when there are model conflicts in the JAXBContext

Sgitario opened the pull request:
When having these two classes:

```java
package io.quarkus.resteasy.reactive.jaxb.deployment.test.one;

import jakarta.xml.bind.annotation.XmlRootElement;

@XmlRootElement
public class Model {
    private String name;
    // ...
}
```

And:
```java
package io.quarkus.resteasy.reactive.jaxb.deployment.test.two;

import jakarta.xml.bind.annotation.XmlRootElement;

@XmlRootElement
public class Model {
    private String name;
    // ...
}
```

If users use the `quarkus-jaxb` extension and they inject the JAXBContext, for example:

```java
@Inject
JAXBContext context;
```

This will fail at runtime because JAXB sees the Model classes as one because of the name. 

These changes address this issue by failing at build time, so users can properly address it by fixing the JAXB model. 

Moreover, users can now exclude any of these classes by using the new property `quarkus.jaxb.exclude-classes`, for example:

```
quarkus.jaxb.exclude-classes=io.quarkus.jaxb.deployment.two.Model
```

Plus, in resteasy reactive, if users exclude a class that is required to serialize/deserialize using JAXB, it will not cause any harm because internally it will recreate a cached JAXBContext including the excluded cache.

Fix https://github.com/quarkusio/quarkus/issues/31032

Sgitario commented:
@Postremus I decided to add logic to detect classes in collision instead of simply not binding the classes found in the endpoints by Resteasy Reactive. I did this for performance reasons and for correctness in the point of view of the `quarkus-jaxb` extension. (As you can reproduce this issue even without the resteasy reactive extension).

gastaldi commented on extensions/jaxb/runtime/src/main/java/io/quarkus/jaxb/runtime/JaxbContextConfigRecorder.java:
```diff
@@ -14,7 +14,15 @@ public void addClassesToBeBound(Collection<String> additionalClassesToBeBound) {
     }
 
+    public static void removeClassesToBeBound(Collection<Class<?>> removeClasses) {
+        removeClasses.forEach(removeClass -> classesToBeBound.removeIf(c -> c.equals(removeClass.getName())));
```
I'd recommend to avoid lambdas in here

Sgitario replied:
Removed

gsmet commented:
Let's wait for @Postremus ' feedback on this one.

geoand commented:
> Let's wait for @Postremus ' feedback on this one.

+1

Postremus commented:
Thank you @Sgitario for working on this!

I got some rudimentary testing done.
The message
```
2023-02-09 21:04:07,816 WARN  [io.qua.jax.run.JaxbContextProducer] (executor-thread-0) The following classes won't be bound to the JAXB context because have the same name 'Root': [org.acme.model1.Root, org.acme.model2.Root]. To resolve these collisions, you need to annotate the class with the `@XmlType` annotation.

```
Is only logged after I invoke the endpoint, and only logged on the first invocation. It could therefore be quickly missed right now.

Maybe it is better to fail server startup? At runtime, all customizer are known.
 The endpoint won't work as defined anyway. At least the message could be printed on startup.


Also there might still be a problem with the collision detection.
1. Why do I need to use XMLType to fix the collision? On my Root Elements I have an `@XMLRootElement` annotation, I believe it should be enough to set the name there like `@XmlRootElement(name = "Root2")`.

2. Collisions are only detected on the global elements, but not on the  local elements.
If I set the `XMLType` on my roots, I still get the following exception:
```
2023-02-09 21:15:15,959 ERROR [org.jbo.res.rea.com.cor.AbstractResteasyReactiveContext] (executor-thread-0) Request failed: java.lang.RuntimeException: org.glassfish.jaxb.runtime.v2.runtime.IllegalAnnotationsException: 1 counts of IllegalAnnotationExceptions
Zwei Klassen haben denselben XML-Typnamen"enum". Verwenden Sie @XmlType.name und @XmlType.namespace, um ihnen unterschiedliche Namen zuzuweisen.                                                                                                                                      
        this problem is related to the following location:                                                                                                                                                                                                                            
                at org.acme.model2.Enum
                at public org.acme.model2.Enum org.acme.model2.Root.getEnum2()
                at org.acme.model2.Root
        this problem is related to the following location:
                at org.acme.model1.Enum
                at public org.acme.model1.Enum org.acme.model1.Root.getEnum1()
                at org.acme.model1.Root

        at io.quarkus.jaxb.runtime.JaxbContextProducer.createJAXBContext(JaxbContextProducer.java:88)
        at io.quarkus.jaxb.runtime.JaxbContextProducer.jaxbContext(JaxbContextProducer.java:35)
        at io.quarkus.jaxb.runtime.JaxbContextProducer_ProducerMethod_jaxbContext_6a6d20272304edd64c7bdfa35e3ed5d5971a3949_Bean.doCreate(Unknown Source)
        at io.quarkus.jaxb.runtime.JaxbContextProducer_ProducerMethod_jaxbContext_6a6d20272304edd64c7bdfa35e3ed5d5971a3949_Bean.create(Unknown Source)
        at io.quarkus.jaxb.runtime.JaxbContextProducer_ProducerMethod_jaxbContext_6a6d20272304edd64c7bdfa35e3ed5d5971a3949_Bean.create(Unknown Source)
        at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:113)
        at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:37)
        at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:34)
        at io.quarkus.arc.impl.LazyValue.get(LazyValue.java:26)
        at io.quarkus.arc.impl.ComputingCache.computeIfAbsent(ComputingCache.java:69)
        at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:34)
        at io.quarkus.jaxb.runtime.JaxbContextProducer_ProducerMethod_jaxbContext_6a6d20272304edd64c7bdfa35e3ed5d5971a3949_Bean.get(Unknown Source)
        at io.quarkus.jaxb.runtime.JaxbContextProducer_ProducerMethod_jaxbContext_6a6d20272304edd64c7bdfa35e3ed5d5971a3949_Bean.get(Unknown Source)
        at io.quarkus.jaxb.runtime.JaxbContextProducer_ProducerMethod_unmarshaller_648fed42516a31dff36335d61646e1afbd13a9fa_Bean.doCreate(Unknown Source)
        at io.quarkus.jaxb.runtime.JaxbContextProducer_ProducerMethod_unmarshaller_648fed42516a31dff36335d61646e1afbd13a9fa_Bean.create(Unknown Source)
        at io.quarkus.jaxb.runtime.JaxbContextProducer_ProducerMethod_unmarshaller_648fed42516a31dff36335d61646e1afbd13a9fa_Bean.create(Unknown Source)
        at io.quarkus.arc.impl.RequestContext.getIfActive(RequestContext.java:74)
        at io.quarkus.arc.impl.ClientProxies.getDelegate(ClientProxies.java:30)
        at jakarta.xml.bind.JaxbContextProducer_ProducerMethod_unmarshaller_648fed42516a31dff36335d61646e1afbd13a9fa_ClientProxy.arc$delegate(Unknown Source)
        at jakarta.xml.bind.JaxbContextProducer_ProducerMethod_unmarshaller_648fed42516a31dff36335d61646e1afbd13a9fa_ClientProxy.unmarshal(Unknown Source)
        at io.quarkus.resteasy.reactive.jaxb.runtime.serialisers.ServerJaxbMessageBodyReader.unmarshal(ServerJaxbMessageBodyReader.java:74)
        at io.quarkus.resteasy.reactive.jaxb.runtime.serialisers.ServerJaxbMessageBodyReader.doReadFrom(ServerJaxbMessageBodyReader.java:96)
        at io.quarkus.resteasy.reactive.jaxb.runtime.serialisers.ServerJaxbMessageBodyReader.readFrom(ServerJaxbMessageBodyReader.java:45)
        at org.jboss.resteasy.reactive.server.handlers.RequestDeserializeHandler.readFrom(RequestDeserializeHandler.java:122)
        at org.jboss.resteasy.reactive.server.handlers.RequestDeserializeHandler.handle(RequestDeserializeHandler.java:81)
        at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:108)
        at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:145)
        at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:576)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
        at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
        at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: org.glassfish.jaxb.runtime.v2.runtime.IllegalAnnotationsException: 1 counts of IllegalAnnotationExceptions
Zwei Klassen haben denselben XML-Typnamen"enum". Verwenden Sie @XmlType.name und @XmlType.namespace, um ihnen unterschiedliche Namen zuzuweisen.
        this problem is related to the following location:
                at org.acme.model2.Enum
                at public org.acme.model2.Enum org.acme.model2.Root.getEnum2()
                at org.acme.model2.Root
        this problem is related to the following location:
                at org.acme.model1.Enum
                at public org.acme.model1.Enum org.acme.model1.Root.getEnum1()
                at org.acme.model1.Root

        at org.glassfish.jaxb.runtime.v2.runtime.IllegalAnnotationsException$Builder.check(IllegalAnnotationsException.java:83)
        at org.glassfish.jaxb.runtime.v2.runtime.JAXBContextImpl.getTypeInfoSet(JAXBContextImpl.java:421)
        at org.glassfish.jaxb.runtime.v2.runtime.JAXBContextImpl.<init>(JAXBContextImpl.java:255)
        at org.glassfish.jaxb.runtime.v2.runtime.JAXBContextImpl$JAXBContextBuilder.build(JAXBContextImpl.java:1115)
        at org.glassfish.jaxb.runtime.v2.ContextFactory.createContext(ContextFactory.java:144)
        at org.glassfish.jaxb.runtime.v2.JAXBContextFactory.createContext(JAXBContextFactory.java:44)
        at jakarta.xml.bind.ContextFinder.find(ContextFinder.java:368)
        at jakarta.xml.bind.JAXBContext.newInstance(JAXBContext.java:605)
        at io.quarkus.jaxb.runtime.JaxbContextProducer.createJAXBContext(JaxbContextProducer.java:86)
        ... 33 more
```

Simply because my `model1.Enum` and `model2.Enum` classes do not have an distinct xml name.

Again, thank you very much <3.

Postremus commented:
I continued testing. I figured out that I do not need XmlType on my roots, I only need it on my `Enum` classes and the problem is resolved.
However, In that case the warning message is wrong.

It says I would need to adjust my Root class and not my Enum class.

Sgitario commented:
@Postremus , I've just updated the pull request based on your feedback.
To sum up:
- We are going to find collisions at build time. (at first, it was at runtime). So the warning message will appear when building your application). Thanks to this, we will have better performance at the startup.

- The collisions will only apply to the types annotated with `@XmlRootElement` or the ones used by RESTEasy Reactive. For these, we'll exclude the types that cause collisions, but RESTEasy Reactive will continue working as it will create a custom JAXBContext instance per type. -> Therefore, we will have the same behavior as in RESTEasy Classic. 

- We won't detect collisions by sub types (for ex: your Enum type). I was trying to implement this logic, but it was getting more complex than expected and I don't think it was worth it. Instead, the build will fail in this use case with a proper message coming from JAXB. 

- I've also updated the warning message to let users they can also use `@XmlRootElement`. 

About the `@XmlRootElement` and `@XmlType` annotations, for root types, we can use either, but if the `@XmlType` annotation is present, this will be used. For non-root types, only the `@XmlType` annotation can be used. So, the logic has been updated according to this. 

Finally, @Postremus I tried again your pull request with these latest changes, and without changing anything, everything will work as expected because both Root classes are going to be excluded. If you resolve the conflict in any of them (by using for example `@XmlRootElement(name = "root1")`, then it will fail at build time because of the Enum conflict which is expected and you will need to resolve yourself by adding the `@XmlType` annotation in any of the Enum types.

Postremus commented:
I intend to do another round of testing and feedback this evening, please hold off merging this if possible. Thanks. :)

gsmet commented:
@Postremus take your time, we are not in a hurry.

Postremus commented:
I found 2 small things that might be worth looking at, otherwise this looks good.

1. @XMLSchema is not supported anymore. This annotation allows to change the namespace of all classes in a package. See for an example https://stackoverflow.com/questions/6895486/jaxb-need-namespace-prefix-to-all-the-elements. When I set this annotation with 3.0.0.Alpha4 in the model2 Package, I can marshall / unmarshall both of my Roots successfully. No need to change the namespace on every class then.
With this pr, the annotation is not respected during collision detection.
I believe camel uses that annotation? Might be worth checking with the camel team.
2. I am not totally happy with the current warning:
```
2023-02-19 20:21:18,184 WARN  [io.qua.jax.dep.JaxbProcessor] (build-29) The following classes won't be bound to the JAXB context because the type match with each other: [org.acme.model2.Root, org.acme.model1.Root]. To resolve these collisions, you need to provide different values on the `name` or `namespace` fields using the `@XmlRootElement` or the `@XmlType` annotations into location that causes this conflict.
```
Yes, the classes are not bound to the default context. However if your endpoints are working just fine. This message is only relevant if you ever intend to directly use the JaxbContext. This could maybe be clarified, since this message is just scary and confusing.

Sgitario commented:
Thanks again for your feedback. 
About 1., taking into account this annotation is too complex and it does not worth it for the problem it resolves. 
About 2., maybe it's better idea to change the warning to debug. 

Anyway, if the current collision detection is faulty and to complex to implement, maybe we should close this pull request and simply document that users must use an unique model according to JAXB in resteasy reactive.

Sgitario commented:
I've updated this pull request to remove the collisions logic, but I wanted to keep the rest of the changes:
- detect JAXB failures at build time
- keep the creation of the JAXB context in RESTEasy reactive when the class is not bound to the default JAXBContext.

I've updated the title and description.
@Postremus would you make taking another look please?

Postremus commented:
Works as advertised.

Thank you for working on this @Sgitario <3

gastaldi commented on docs/src/main/asciidoc/resteasy-reactive.adoc:
```diff
@@ -1353,6 +1353,15 @@ Importing this module will allow HTTP message bodies to be read from XML
 and serialised to XML, for <<resource-types,all the types not already registered with a more specific
 serialisation>>.
 
+The JAXB resteasy reactive extension will automatically detect the classes that are used in the resources and require of a JAXB serialization. Then, it will register these classes into the default JAXBContext which is internally used by the JAXB message reader and writer. 
```
```suggestion
The JAXB Resteasy Reactive extension will automatically detect the classes that are used in the resources and require JAXB serialization. Then, it will register these classes into the default `JAXBContext` which is internally used by the JAXB message reader and writer. 
```

gastaldi commented on docs/src/main/asciidoc/resteasy-reactive.adoc:
```diff
@@ -1353,6 +1353,15 @@ Importing this module will allow HTTP message bodies to be read from XML
 
+The JAXB resteasy reactive extension will automatically detect the classes that are used in the resources and require of a JAXB serialization. Then, it will register these classes into the default JAXBContext which is internally used by the JAXB message reader and writer. 
+
+However, in some situations, these classes cause the JAXBContext to fail: for example, when you're using the same class name in different java packages. In these cases, the application will fail at build time and print the JAXB exception that caused the issue, so you can properly fix it. Alternatively, you can also exclude the classes that cause the issue by using the property `quarkus.jaxb.exclu ...
```
```suggestion
However, in some situations, these classes cause the `JAXBContext` to fail: for example, when you're using the same class name in different java packages. In these cases, the application will fail at build time and print the JAXB exception that caused the issue, so you can properly fix it. Alternatively, you can also exclude the classes that cause the issue by using the property `quarkus.jaxb.exclude-classes`. When excluding classes that are required by any resource, the JAXB resteasy reactive extension will create and cache a custom `JAXBContext` that will include the excluded class, causing a minimal performance degradance. 
```

gastaldi commented on docs/src/main/asciidoc/resteasy-reactive.adoc:
```diff
@@ -1353,6 +1353,15 @@ Importing this module will allow HTTP message bodies to be read from XML
+
+[NOTE]
+====
+The property `quarkus.jaxb.exclude-classes` accepts a comma separated list of full qualified class names, for example: `quarkus.jaxb.exclude-classes=org.acme.one.Model,org.acme.two.Model`. 
```
```suggestion
The property `quarkus.jaxb.exclude-classes` accepts a comma separated list of fully qualified class names, for example: `quarkus.jaxb.exclude-classes=org.acme.one.Model,org.acme.two.Model`. 
```

Postremus commented:
@Sgitario Do you think this is safe to backport?

Sgitario commented:
> @Sgitario Do you think this is safe to backport?

We should not backport this as it brings a possible breaking change for users who had an invalid JAXB Context (it would fail for them at build time). I need to list this change in the migration guide too.

Sgitario committed:
Fail at build time when there are model conflicts in the JAXBContext.

When having these two classes:

```java
package io.quarkus.resteasy.reactive.jaxb.deployment.test.one;

import jakarta.xml.bind.annotation.XmlRootElement;

@XmlRootElement
public class Model {
    private String name;
    // ...
}
```

And:
```java
package io.quarkus.resteasy.reactive.jaxb.deployment.test.two;

import jakarta.xml.bind.annotation.XmlRootElement;

@XmlRootElement
public class Model {
    private String name;
    // ...
}
```

If users use the `quarkus-jaxb` extension and they inject the JAXBContext, for example:

```java
@Inject
JAXBContext context;
```

This will fail at runtime because JAXB sees the Model classes as one because of the name. 

These changes address this issue by failing at build time, so users can properly address it by fixing the JAXB model. 

Moreover, users can now exclude any of these classes by using the new property `quarkus.jaxb.exclude-classes`, for example:

```
quarkus.jaxb.exclude-classes=io.quarkus.jaxb.deployment.two.Model
```

Plus, in resteasy reactive, if users exclude a class that is required to serialize/deserialize using JAXB, it will not cause any harm because internally it will recreate a cached JAXBContext including the excluded cache.

Sgitario commented:
Added note in the migration guide: https://github.com/quarkusio/quarkus/wiki/Migration-Guide-3.0#jaxb

DeMol-EE commented:
I’m having a little issue with my project in the sense that I have over 3000 classes that I would need to exclude. I tried putting them in one very long config property value, but that leads to the following error:

```
[error]: Build step io.quarkus.deployment.steps.ConfigGenerationBuildStep#buildTimeRunTimeConfig threw an exception: java.lang.IllegalArgumentException: Cannot load strings larger than 65535 bytes
```

Could I find any more documentation on this new config property, or is there perhaps a way to disable the build time check all together? The thing with my project is that it needs to support 2 different versions of the same jaxb classes. They are kept in 2 different packages. For each request, the code checks which context it should use and initialized a context instance programmatically if one doesn’t exist yet (with `JAXBContext.newInstance(...)`). The was working fine on Quarkus 2 because the code does not rely on CDI for injecting a context, but this new build time check is preventing me from building. What should I do?

Sgitario commented:
> I’m having a little issue with my project in the sense that I have over 3000 classes that I would need to exclude. I tried putting them in one very long config property value, but that leads to the following error:
> 
> ```
> [error]: Build step io.quarkus.deployment.steps.ConfigGenerationBuildStep#buildTimeRunTimeConfig threw an exception: java.lang.IllegalArgumentException: Cannot load strings larger than 65535 bytes
> ```
> 
> Could I find any more documentation on this new config property, or is there perhaps a way to disable the build time check all together? The thing with my project is that it needs to support 2 different versions of the same jaxb classes. They are kept in 2 different packages. For each request, the code checks which context it should use and initialized a context instance programmatically if one doesn’t exist yet (with `JAXBContext.newInstance(...)`). The was working fine on Quarkus 2 because the code does not rely on CDI for injecting a context, but this new build time check is preventing me from building. What should I do?

From Quarkus 3.0.0.Alpha5, you can disable the validation of the JAXBContext instance by using `quarkus.jaxb.validate-jaxb-context=false`.

And from Quarkus 3.0.0.Beta1, you can exclude packages using the existing property `quarkus.jaxb.exclude-classes`. For example: `quarkus.jaxb.exclude-classes=org.package.*`.

gsmet commented:
@Sgitario I'm more and more thinking that this `@Inject JAXBContext` feature is coming with more issues than it actually solves.
Did we have users asking for the ability to inject a preconfigured `JAXBContext` with all the JAXB classes around in it?

Sgitario commented:
> @Sgitario I'm more and more thinking that this `@Inject JAXBContext` feature is coming with more issues than it actually solves. Did we have users asking for the ability to inject a preconfigured `JAXBContext` with all the JAXB classes around in it?

And I'm more more thinking that you're right :)
This was actually a request from one user, but perhaps we could disable it by default.

geoand commented:
+1 for disabling by default

Sgitario commented:
FYI: https://github.com/quarkusio/quarkus/pull/32105
